diffprivlib/mechanisms/snapping.py

```diff
@@ -0,0 +1,232 @@
+
+"""
+The Snapping mechanism in differential privacy, which eliminates a weakness to floating point errors in the classic
+Laplace mechanism with standard Laplace sampling.
+"""
+import math
+import struct
+
+import crlibm
+import numpy as np
+
+from diffprivlib.mechanisms import LaplaceTruncated
+
+
+class Snapping(LaplaceTruncated):
+    r"""
+    The Snapping mechanism for differential privacy.
+
+    First proposed by Ilya Mironov [M12]_.
+
+    It eliminates a vulnerability stemming from the representation of reals as floating-point numbers in implementations
+    of the classic Laplace mechanism and its variants which use the inverse CDF of the Laplace distribution to sample
+    it. It causes a high degree of reduction in the granularity of the output.
+
+    Parameters
+    ----------
+    epsilon : float
+        Privacy parameter :math:`\epsilon` for the mechanism.  Must be in [:math:`2*\eta`, ∞], where math:`\eta` is the
+        machine epsilon of the floating point type.
+
+    sensitivity : float
+        The sensitivity of the mechanism.  Must be in [0, ∞).
+
+    lower : float
+        The lower bound of the mechanism.
+
+    upper : float
+        The upper bound of the mechanism.
+    References
+    ----------
+    .. [Mir12] Mironov, Ilya. "On significance of the least significant bits for differential privacy." Proceedings of
+     the 2012 ACM conference on Computer and communications security (2012).
+    """
+    def __init__(self, *, epsilon, sensitivity, lower, upper):
+        super().__init__(epsilon=epsilon, sensitivity=sensitivity, delta=0.0, lower=lower, upper=upper)
+        self.epsilon = self._check_epsilon_machine_epsilon(epsilon)
+        self._bound = self._scale_bound()
+        self._epsilon_0 = self.effective_epsilon()
```

Comment:
Can we do away with defining `_epsilon_0` altogether to avoid cluttering the namespace, and just use `self.effective_epsilon()` instead? Seeing as `_epsilon_0` isn't used very much, this won't add much code clutter.